daml-script/daml3/Daml/Script/Questions/Query.daml

```diff
@@ -47,15 +47,21 @@ instance IsQuestion QueryContractId (Optional AnyTemplate) where command = "Quer
 --
 -- This is semantically equivalent to calling `query`
 -- and filtering on the client side.
-queryContractId : forall t p. (Template t, HasAgreement t, IsParties p) => HasCallStack => p -> ContractId t -> Script (Optional t)
+queryContractId_ : forall t p. (Template t, IsParties p) => HasCallStack => p -> ContractId t -> Script (Optional (AnyTemplate, Text))
 -- The 'HasAgreement t' constraint prevents this function from being used on interface types.
-queryContractId p c = fmap convert $ lift $ QueryContractId with
+queryContractId_ p c = lift $ QueryContractId with
     parties = toParties p
     tplId = templateTypeRep @t
     cid = coerceContractId c
-  where
-    convert : Optional AnyTemplate -> Optional t
-    convert = fmap $ fromSome . fromAnyTemplate
+--  where
+--    convert : Optional AnyTemplate -> Optional t
+--    convert = fmap $ fromSome . fromAnyTemplate
+
+queryContractId: forall t p. (Template t, HasAgreement t, IsParties p) => HasCallStack => p -> ContractId t -> Script (Optional t)
+queryContractId p c = fmap (fmap $ fromSome . fromAnyTemplate . fst) $ queryContractId_ p c
+
+queryDisclosureContractId: forall t p. (Template t, IsParties p) => HasCallStack => p -> ContractId t -> Script (Optional Disclosure)
+queryDisclosureContractId p c = fmap (fmap  $ Disclosure . snd) $ queryContractId_ p c
```

Comment:
Are are assuming someone would never want both?
I'd probably move the wrapping logic for each into a single function that gives `Optional (t, Disclosure)` then make these simple functions to peek into the data

Reply:
Nothing against the idea. 
Just wondering if we will not better redesign all this mess before. (Eg. The names of all those fucntion are pretty messy) 
For now we have implementation you want is just not exposed (`queryContractId_`)

Reply:
Okay, I'm happy if you can put up a quick ticket to redo this before daml3